source/connection-monitoring-and-pooling/tests/pool-acquire.yml

```diff
@@ -0,0 +1,147 @@
+version: 1
+style: unit
+tests:
+  - description: must be able to acquire a connection
+    operations:
+      - name: createPool
+        returnTo: pool
+      - name: acquire
+        object: pool
+    events:
+      - type: connectionPoolQueueEntered
+        pool:
+          id: {$$ref: ['pool', 'id']}
+      - type: connectionPoolAcquire
+        pool:
+          id: {$$ref: ['pool', 'id']}
+        connection:
+          id: 0
+    ignore:
+      - connectionCreated
+      - connectionPoolCreated
+  - description: must be able to acquire multiple connections at the same time
+    operations:
+      - name: createPool
+        returnTo: pool
+      - name: start
+        args: [ thread1 ]
+      - name: start
+        args: [ thread2 ]
+      - name: start
+        args: [ thread3 ]
+      - name: acquire
+        thread: thread1
+        object: pool
+        returnTo: conn1
+      - name: acquire
+        thread: thread2
+        object: pool
+        returnTo: conn2
+      - name: acquire
+        thread: thread3
+        object: pool
+        returnTo: conn3
+      - name: waitFor
+        args: [ thread1 ]
+      - name: waitFor
+        args: [ thread2 ]
+      - name: waitFor
+        args: [ thread3 ]
+    events:
+      - type: connectionPoolAcquire
+        pool:
+          id: {$$ref: [ pool, id ]}
+        connection:
+          id: 0
+      - type: connectionPoolAcquire
+        pool:
+          id: {$$ref: [ pool, id ]}
+        connection:
+          id: 1
+      - type: connectionPoolAcquire
+        pool:
+          id: {$$ref: [ pool, id ]}
+        connection:
+          id: 2
```

Comment:
Since connection creation (which defines the id) and connection return from the pool are separate operations, there is no requirement that the events for connection checkout will have the ids in ascending order.

Reply:
The pool is created with no minPoolSize, meaning there are no pre-generated connections. So every call to acquire will result in the creation of a connection for that call. The calls to acquire are in parallel, but the waitQueue mandates that they run acquisition one at a time, ensuring that each one will get a connection in order.

Reply:
The spec does not say that connection creation must be performed under the lock of a wait queue, only that the wait queue must be threadsafe  and fair. If connections are (always) created by a background thread, creating them under the wait queue's lock is at a minimum very difficult. If connection timeout exceeds wait queue timeout, a connection won't ever be created by a client so configured. Even if the tests don't perform any network i/o there is still a lot that happens between when an id is allocated and when a connection can be checked out, and if these operations are done in a background thread then the sequence of ids as seen by the application may be non-monotonic even in the absence of network i/o.

Reply:
If there is a single background thread that creates all connections, then, in the absence of network i/o, I imagine the sequence of connection ids as exposed to the application through events should in fact be monotonic. But once network i/o gets added to the picture, even async i/o on the same thread, the sequence would no longer need to be monotonic (e.g. due to handshake taking varying amounts of time and thus connections becoming ready in a different order from their creation/id allocation).